typing/includecore.ml

```diff
@@ -127,17 +137,27 @@ let value_descriptions ~loc env name
       | Ok () -> ()
       | Error e -> raise (Dont_match (Modality e))
       end;
-  | Legacy when (vd1.val_modalities == vd2.val_modalities) ->
+  | Legacy close_over_coercion ->
+    match Mode.Modality.Value.to_const_opt vd2.val_modalities with
       (* [wrap_constraint_with_shape] invokes inclusion check with identical
         inferred modalities, which we need to workaround. *)
-      ()
-  | Legacy ->
+    | None -> ()
+    | Some val2_modalities ->
       let mmode1, mmode2 =
         Mode.Value.(disallow_right legacy), Mode.Value.(disallow_left legacy)
       in
       let mode1 = Mode.Modality.Value.apply vd1.val_modalities mmode1 in
-      let mode2 =
-        Mode.Modality.Value.(Const.apply (to_const_exn vd2.val_modalities) mmode2)
+      let mode2 = Mode.Modality.Value.(Const.apply val2_modalities mmode2) in
+      let mode1 =
+        match close_over_coercion with
+        | Some held_locks ->
+          (* Similiar to [Typecore.type_ident] *)
+          let mode1 = left_mode_cross env vd2.val_type mode1 in
```

Comment:
Can you give an intuition for why `vd2.val_type` is used here rather than `vd1.val_type`?